### PR TITLE
Create Blog “open-sourcing-packetrusher-a-5g-core-performance-tester”

### PR DESCRIPTION
**Pull request by linouxis9**
Automatically generated by Netlify CMS

Still WIP, but I wanted to check the preview :-)
Thanks a lot!

### CODE REVIEW

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+In the fast-evolving landscape of 5G technology, the demand for robust and efficient testing tools has never been higher. Enter PacketRusher, a cutting-edge 5G Core Network performance testing tool. As the lead developer behind this groundbreaking project, I am thrilled to share the power and potential PacketRusher holds in revolutionizing the way we test and optimize 5G networks.
+
+## PacketRusher
+
+PacketRusher is a tool dedicated to the performance testing and automatic validation of 5G Core Networks. It tests a 5G Core Network via its external interfaces, as defined by the 3GPP : N1, N2 and N3 interfaces. The tool can emulate various types of User Equipment (eg. 5G phones or 5G-enabled IoT devices) and gNodeBs (eg. antennas in Radio Access Network), and it can validate and load-tests the 5G core network’s Control Plane functions using a set of pre-defined procedures on top of both New Generation Application Protocol (NGAP) and Non-access Stratum (NAS-5GS) 5G protocols. It enables users to tests 5G Core Networks with up to 100k simulated UEs simultaneously all from a simple Linux virtual machine, without the needs for expensive commercial tools, or 100k phones on a table ;-)
```
Comment (choukden) [inline content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:19]: Suggest to define 3GPP in the second sentence:
...as defined by the 3rd Generation Partnership Project (3GPP): N1, N2 and N3 interfaces.

You can also add a note about what are the N1, N2 and N3 interfaces.

Remember, our readers are not all experts in 5G cell networking. 

Last sentence:
...with up to 100k simulated User Equipment (UE) simultaneously...

Equipment has no plural form.

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+To summarize, there are essentially three main components:
+- User Equipment, which is any 5G device used directly by a user to communicate. For example, a smartphone or an IoT device.
+- The gNodeB, which is the radio used for wireless communication between the User Equipment and the 5G network core.
+- And finally, the 5G network core itself, which manages all mobile network functionalities. This is the component that will manage phone authentication in the network, quality of service, routing to the data network, and so on...
+
+
+![﻿High-level diagram of the PacketRusher's architecture and its interaction with a 5G Core's AMF and UPF](/img/packetrusher-architecture.png "﻿High-level diagram of the PacketRusher's architecture and its interaction with a 5G Core's AMF and UPF")
+
+PacketRusher's behavior is quite simple; it essentially simulates two of the three major components of a 5G network: 
+- the User Equipments, i.e. the phones
```
Comment (choukden) [inline content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:35]: If you look at the preview here: https://deploy-preview-2223--hpe-dev-portal.netlify.app/blog/open-sourcing-packetrusher-a-5g-core-performance-tester/

the bullet list of items is not formatted correctly. you can use:

* The User Equipment, i.e. the phones.
* the gNodesB, i.e. the antennas.

At the same time, it connect to the 5G Core via....

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+## PacketRusher
+
+PacketRusher is a tool dedicated to the performance testing and automatic validation of 5G Core Networks. It tests a 5G Core Network via its external interfaces, as defined by the 3GPP : N1, N2 and N3 interfaces. The tool can emulate various types of User Equipment (eg. 5G phones or 5G-enabled IoT devices) and gNodeBs (eg. antennas in Radio Access Network), and it can validate and load-tests the 5G core network’s Control Plane functions using a set of pre-defined procedures on top of both New Generation Application Protocol (NGAP) and Non-access Stratum (NAS-5GS) 5G protocols. It enables users to tests 5G Core Networks with up to 100k simulated UEs simultaneously all from a simple Linux virtual machine, without the needs for expensive commercial tools, or 100k phones on a table ;-)
+
+## Architecture
+Before we understand how PacketRusher works, we need to understand the components of a minimal 5G network:
+
+![High-level diagram of a 5G deployment](/img/high-level-diagram-of-a-5g-deployment.png "High-level diagram of a 5G deployment")
+
+To summarize, there are essentially three main components:
```
Comment (choukden) [inline content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:26]: If you look at the preview here: https://deploy-preview-2223--hpe-dev-portal.netlify.app/blog/open-sourcing-packetrusher-a-5g-core-performance-tester/

the bullet list of items is not formatted correctly. you can use:

To summarize, there are essentially three main components:

* User Equipment, which is....
* The gNodeB, whihc is the radio...
* The 5G network core itself, which manages...

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+![﻿High-level diagram of the PacketRusher's architecture and its interaction with a 5G Core's AMF and UPF](/img/packetrusher-architecture.png "﻿High-level diagram of the PacketRusher's architecture and its interaction with a 5G Core's AMF and UPF")
+
+PacketRusher's behavior is quite simple; it essentially simulates two of the three major components of a 5G network: 
+- the User Equipments, i.e. the phones
+- and the gNodeB, i.e. the antennas.
+- At the same time, it connects to the 5G Core via its external interfaces, as if it were a black box.
+PacketRusher simulates both components en masse, which means it can simulate several user devices and gNodeBs at the same time.
+
+For testing, PacketRusher implements main control plane procedures:
```
Comment (choukden) [inline content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:40]: Bullet item list are not formatted correctly.
Make sure to include a \<space\> between the * and the text.

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  * Supports UE attach/detach (registration/authentifcation/security mode) procedures
+  * Supports Create/Delete PDU Sessions,  up to 15 PDU Sessions per UE
+  * Supports Xn handover: UE handover between simulated gNodeB (PathSwitchRequest)
+  * Supports 5G Roaming
+
+For throughput testing, PacketRusher implements high-performant N3 (GTP-U) interface:
```
Comment (choukden) [inline content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:46]: Same remark for the bullet list here.

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  * Generic tunnel supporting all kind of traffic (TCP, UDP, Video…)
+    * We tested iperf3 traffic, and Youtube traffic through PacketRusher.
+    * We roughly reach 5 GB/s per UE, which is more than what a real UE can achieve.
+
+
+## Community contributions
+
+Excitingly, PacketRusher has caught the attention of a leading industry player. Orange, a key player in the telecommunications sector, has recognized the potential of PacketRusher and is actively integrating it into their open-source GitHub project, [toward-5gs](https://github.com/Orange-OpenSource/towards5gs-helm).\
```
Comment (choukden) [inline content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:54]: Remove the backslack (\\) at the end of the sentence.

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+We can also mention the outstanding work of Github User [s5uishida](https://github.com/s5uishida) who made a [high-quality performance comparison](https://github.com/s5uishida/simple_measurement_of_upf_performance) of open-source UPF using PacketRusher.
+
+
+## Join the 5G Revolution
+
+We believe that the future of 5G is a collaborative one, and we invite you to be a part of this exciting journey. Whether you are a seasoned developer, a researcher, or simply passionate about the possibilities of 5G, your contributions can make a significant impact.\
```
Comment (choukden) [inline content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:60]: Remove the backslash (\\) at the end of the sentence.

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+To get started, visit our [GitHub repository](https://github.com/HewlettPackard/PacketRusher) to access the source code, documentation, and engage with the community. Your feedback, suggestions, and contributions are invaluable in shaping the future of 5G technology.\
```
Comment (choukden) [inline content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:61]: Remove the backslash (\\) at the end of the sentence.

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+At HPE, we are committed to pushing the boundaries of innovation, and with the release of PacketRusher as open-source, we are laying the foundation for a new era of connectivity.   
+
+PacketRusher is not just a testing tool; it's a community-driven effort to push the boundaries of 5G technology. Let's build the future of 5G together!
+
+[V﻿alentin D'Emmanuele](https://www.linkedin.com/in/valentin-d-emmanuele/),\
```
Comment (choukden) [inline content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:66]: Remove the backslash (\\) at the end of the sentence.

**Review by choukden — CHANGES_REQUESTED** [review]
Hi Valentin, 

Thank you very much for your great contribution to the HPE Dev Community about the open source project PacketRusher. Really a very interesting subject.

I have put a suggestions for edits. Please also check how to format bullet list item.
I also invite you to look at the Tips #16 in the CMS guide: https://github.com/hpe-dev-incubator/hpe-dev-portal/blob/master/docs/ContributorGuide-v2.md#tips-and-trick-using-the-cms-editor.

Dale Rensing, our editorial consultant, will go through the editorial review in the coming days or so. Expect to receive additional grammatical/editorial suggestions for changes to incorporate in the blog post before we publish it to the HPE Dev Community portal.

To review the suggestions and acknowledge them in GitHub:

- Navigate to **File Changed** area to visualize the comments from start to the end, or navigate to **Conversation** area to visualize the comments per reviewer.

- You can mark the suggestion you have incorporated in your platform page entry in the CMS Workflow Editor by clicking on button **Resolve Conversation.**

Note we recommend you continue doing edits of your platform entry in the CMS Workflow editor while the entry is in the "**In Review**" column. Moving it back to the "**Drafts**" column would delete the Pull Request in Github.

Once you have completed making your edits, please send an email to [denis.choukroun@hpe.com](mailto:denis.choukroun@hpe.com) and [dale.rensing@hpe.com](mailto:dale.rensing@hpe.com)  indicating that your blog post is ready for its final review. 

Thank you very much

Denis

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md [hunk 10]
```diff
@@ -0,0 +1,67 @@
+---
+title: "Open-sourcing PacketRusher: A 5G Core performance tester"
```
Comment (drensing) [inline content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:2]: Hi there. I'm your friendly neighborhood HPE Developer blog post editor. I'll be editing your post mostly for grammar, HPE Brand compliance, and readability.

One thing I want to point out from the start is that HPE Brand requests that we use "sentence case" in anything we write, whether it's blog posts, presentations, web pages, etc.  This means that, as opposed the traditional "title case", where you capitalize the initial letter in each word that might be important in a title (example: How to Bake a Cake), instead you only initially capitalize the first word in a sentence, headline or title. The two exceptions to this rule are when the word happens to be a proper noun (i.e. the name or title of something (example: Microsoft Word) OR it's as shown on a screen capture (example: Click Here - if it's actually shown on the screen that way in your demo).

This means, as an editor, I need to learn about all the different proper names in technology. So I go around googling a variety of items that I'm not familiar with. In your case, it's 5G Core. This one is problematic in that, on some pages it's shown as "5G Core" and on others it's shown as "5G core". 

So... as I am not the subject matter expert on this topic, I'm going to leave it to you to decide which way it should be based on the guidance I've given you above :-). I hope that helps.

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+thumbnailimage: /img/thumbnail.png
+disable: false
+tags:
+  - opensource
+  - 5g
+  - gnodeb
+  - ue
+  - simulator
+---
+In the fast-evolving landscape of 5G technology, the demand for robust and efficient testing tools has never been higher. Enter PacketRusher, a cutting-edge 5G Core Network performance testing tool. As the lead developer behind this groundbreaking project, I am thrilled to share the power and potential PacketRusher holds in revolutionizing the way we test and optimize 5G networks.
```
Comment (drensing) [inline content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:15]: I will, however, give my opinion. I think here it would best be written as "5 G core network performance testing tool. But I could easily be persuaded in the other direction, especially since you do seem to be an expert in this particular subject :-).

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md (hunk 1 above)
Comment (drensing) [inline content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:19]: In the second sentence in this paragraph, please remove the comma after "external interfaces".
"User Equipment" should be written as "user equipment" (that one I do know :-) ).
I recommend breaking out the third sentence into two separate sentences:
1) The tool can emulate various types of user equipment.... and gNodeBs (eg. antennas in Radio Access Network).
2) It can also validate and load test the 5G core network's control plane functions using a set of pre-defined...5G protocols.

Otherwise, it would be quite a long sentence to sift through. Please note where I changed "load-tests" to "load test" and "Control Plane" to "control plane".

In the final sentence, change "tests" to "test" so it reads:
"It enables users to test 5G Core Networks with up to...

Also, change "the needs" to "the need".

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+## PacketRusher
+
+PacketRusher is a tool dedicated to the performance testing and automatic validation of 5G Core Networks. It tests a 5G Core Network via its external interfaces, as defined by the 3GPP : N1, N2 and N3 interfaces. The tool can emulate various types of User Equipment (eg. 5G phones or 5G-enabled IoT devices) and gNodeBs (eg. antennas in Radio Access Network), and it can validate and load-tests the 5G core network’s Control Plane functions using a set of pre-defined procedures on top of both New Generation Application Protocol (NGAP) and Non-access Stratum (NAS-5GS) 5G protocols. It enables users to tests 5G Core Networks with up to 100k simulated UEs simultaneously all from a simple Linux virtual machine, without the needs for expensive commercial tools, or 100k phones on a table ;-)
+
+## Architecture
+Before we understand how PacketRusher works, we need to understand the components of a minimal 5G network:
```
Comment (drensing) [inline content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:22]: In the blog posts, it's recommended that one avoids the "royal we" and instead speaks directly to the person reading the blog post. For example, I would change "Before we understand how..., we need to..." this:
"Before I show you how PacketRusher works, it's important to have an understanding of the components of a minimal 5G network."

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+![High-level diagram of a 5G deployment](/img/high-level-diagram-of-a-5g-deployment.png "High-level diagram of a 5G deployment")
+
+To summarize, there are essentially three main components:
+- User Equipment, which is any 5G device used directly by a user to communicate. For example, a smartphone or an IoT device.
```
Comment (drensing) [inline content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:27]: Don't use initial capital letters for "User Equipment". You can state "User equipment" at the beginning of a sentence or bullet, but when it's in the middle of the sentence, it should be "user equipment".

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md (hunk 2 above)
Comment (drensing) [inline content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:35]: Same comment on "User Equipment"

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+
+PacketRusher's behavior is quite simple; it essentially simulates two of the three major components of a 5G network: 
+- the User Equipments, i.e. the phones
+- and the gNodeB, i.e. the antennas.
+- At the same time, it connects to the 5G Core via its external interfaces, as if it were a black box.
+PacketRusher simulates both components en masse, which means it can simulate several user devices and gNodeBs at the same time.
+
+For testing, PacketRusher implements main control plane procedures:
+  * Supports UE attach/detach (registration/authentifcation/security mode) procedures
+  * Supports Create/Delete PDU Sessions,  up to 15 PDU Sessions per UE
```
Comment (drensing) [inline content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:42]: Here it should be "Supports create/delete PDU sessions, up to 15 PDU sessions per EU

By the way, have we defined PDU yet? If not, it makes sense to define it in its first instance of use by spelling it out completely and then writing (PDU) after it. For example, Primary Digital Unit (PDU)... (I know that's probably not what it means, but I don't know what it means... hence my suggestion :-)

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md (hunk 5 above)
Comment (drensing) [inline content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:46]: Change "implements high-performant N3" to "implements a high-performance N3". The use of "performant" would require the adverb "highly" and an article prior to it, so it would need to read "implements a highly-performant N3".

However, I'm not sure the word "implements" is really what you mean here. You can "use" a specific type of interface, but you "implement" a specific methodology or strategy. Do you see what I mean?

This paragraph is a little confusing to me. There appear to be bullet notations, but after those notations are actual sentences, i.e. "We roughly reach 5 GB/s per UE, which is more than what a real EU can achieve." Were you really thinking of making these bullets? It might read much better as a paragraph. Or, at least, the last two sentences might be better off in their own paragraph?  Take another look at the structure of this and see how you might do it differently, with my and Denis comments in mind.

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+For throughput testing, PacketRusher implements high-performant N3 (GTP-U) interface:
+  * Generic tunnel supporting all kind of traffic (TCP, UDP, Video…)
+    * We tested iperf3 traffic, and Youtube traffic through PacketRusher.
+    * We roughly reach 5 GB/s per UE, which is more than what a real UE can achieve.
+
+
+## Community contributions
+
+Excitingly, PacketRusher has caught the attention of a leading industry player. Orange, a key player in the telecommunications sector, has recognized the potential of PacketRusher and is actively integrating it into their open-source GitHub project, [toward-5gs](https://github.com/Orange-OpenSource/towards5gs-helm).\
+We can also mention the outstanding work of Github User [s5uishida](https://github.com/s5uishida) who made a [high-quality performance comparison](https://github.com/s5uishida/simple_measurement_of_upf_performance) of open-source UPF using PacketRusher.
```
Comment (drensing) [inline content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:55]: Again, in the Community contributions section and the Join the 5G Revolution section, try to avoid the "royal we". Simply state "I'd like to also mention the outstanding work of..." here

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+
+
+## Join the 5G Revolution
```
Comment (drensing) [inline content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:58]: Change "Revolution" to "revolution"

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md (hunk 7 above)
Comment (drensing) [inline content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:60]: Change "We believe that the future of 5G is a collaborative one, and we invite you to be..." to the following:
"How 5G evolves will be based on community collaboration and you're invited to be a part of this exciting journey."

Remove the backslash at the end of the sentence here and below.

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+
+We believe that the future of 5G is a collaborative one, and we invite you to be a part of this exciting journey. Whether you are a seasoned developer, a researcher, or simply passionate about the possibilities of 5G, your contributions can make a significant impact.\
+To get started, visit our [GitHub repository](https://github.com/HewlettPackard/PacketRusher) to access the source code, documentation, and engage with the community. Your feedback, suggestions, and contributions are invaluable in shaping the future of 5G technology.\
+At HPE, we are committed to pushing the boundaries of innovation, and with the release of PacketRusher as open-source, we are laying the foundation for a new era of connectivity.
```
Comment (drensing) [inline content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:62]: Because you're talking about HPE here, it's probably okay to use the "royal we" here.

**Review by drensing — COMMENTED** [review]
I hope you've found my comments helpful, if not entertaining.
This is a really great blog post and our team is so excited that you've decided to publish it here. Thank you!!

**Comment by linouxis9** [comment]
Hi @choukden @drensing
Thank you both for the review! It was indeed both helpful and entertaining :-) I'll proceed with the requested changes and revert back to you by email once done.
Thanks again for having me in your blog and sharing some tips on the HPE Brand guidelines!

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md (hunk 10 above)
Comment (linouxis9) [reply content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:2]: Hi @drensing! Thanks a lot for checking that!! And well, the answer to this question is not easy. I think why you observed both "5G Core" and "5G core" is that we can refer to the "5G Core" both with a common noun and a proper noun.
A bit like if "word" were to be both the common noun for a word processor and "Word" the proper noun of a given word processors. I think it's pretty similar here, "5G Core" refers to the central entity to a 5G network, while "5G core" basically mean the core of the 5G network which is of course the "5G Core".
So TL;DR, I would say it's a bit like history that can be used either as a proper noun and a common noun, but I'm not a linguist and I'm far from being a native English-speaker! Thanks for the thoughts experiments in any cases :D

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md (hunk 11 above)
Comment (linouxis9) [reply content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:15]: I agree!

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md (hunk 1 above)
Comment (linouxis9) [reply content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:19]: Will do, thanks!

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md (hunk 5 above)
Comment (linouxis9) [reply content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:46]: Would "implements a highly-performant N3 tunnel" work?

Yes! These bullets are things that I had not yet written as paragraph yet and that I needed to write before we publish the post.

File: content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md (hunk 1 above)
Comment (linouxis9) [reply content/blog/open-sourcing-packetrusher-a-5g-core-performance-tester.md:19]: Will do, thanks! One of the big issues I have when writting texts is that I seem to always write seemingly never-ending sentences or paragraph-sized sentences :-)
